Cubical/HITs/Join/Properties.agda

```diff
@@ -0,0 +1,274 @@
+{-
+
+This file contains:
+
+- Equivalence with the pushout definition
+  Written by: Loïc Pujet, September 2019
+
+- Associativity of the join
+  Written by: Loïc Pujet, September 2019
+
+-}
+
+{-# OPTIONS --cubical --safe #-}
+
+module Cubical.HITs.Join.Properties where
+
+open import Cubical.Core.Glue
+
+open import Cubical.Foundations.Prelude
+open import Cubical.Foundations.Equiv
+open import Cubical.Foundations.Isomorphism
+open import Cubical.Foundations.GroupoidLaws
+
+open import Cubical.Data.Prod
+
+open import Cubical.HITs.Join.Base
+open import Cubical.HITs.Pushout
+
+private
+  variable
+    ℓ : Level
+    ℓ' : Level
+
+
+-- Alternative definition of the join using a pushout
+joinAlt : (A : Type ℓ) → (B : Type ℓ') → Type (ℓ-max ℓ ℓ')
+joinAlt A B = Pushout {_} {_} {_} {A × B} {A} {B} proj₁ proj₂
+
+-- Proof that it is equal
+joinAlt-iso-join : (A : Type ℓ) → (B : Type ℓ') → Iso (joinAlt A B) (join A B)
+joinAlt-iso-join A B = iso (joinAlt→join A B) (join→joinAlt A B) (join→joinAlt→join A B) (joinAlt→join→joinAlt A B)
+  where
+    joinAlt→join : (A : Type ℓ) → (B : Type ℓ') → joinAlt A B → join A B
+    joinAlt→join A B (inl x) = inl x
+    joinAlt→join A B (inr x) = inr x
+    joinAlt→join A B (push y i) = push (proj₁ y) (proj₂ y) i
+
+    join→joinAlt : (A : Type ℓ) → (B : Type ℓ') → join A B → joinAlt A B
+    join→joinAlt A B (inl x) = inl x
+    join→joinAlt A B (inr x) = inr x
+    join→joinAlt A B (push a b i) = push (a , b) i
+
+    joinAlt→join→joinAlt : (A : Type ℓ) → (B : Type ℓ') → ∀ x → join→joinAlt A B (joinAlt→join A B x) ≡ x
+    joinAlt→join→joinAlt A B (inl x) = refl
+    joinAlt→join→joinAlt A B (inr x) = refl
+    joinAlt→join→joinAlt A B (push (a , b) j) = refl
+
+    join→joinAlt→join : (A : Type ℓ) → (B : Type ℓ') → ∀ x → joinAlt→join A B (join→joinAlt A B x) ≡ x
+    join→joinAlt→join A B (inl x) = refl
+    join→joinAlt→join A B (inr x) = refl
+    join→joinAlt→join A B (push a b j) = refl
+
+-- We will need both the equivalence and path version
+joinAlt≃join : (A : Type ℓ) → (B : Type ℓ') → joinAlt A B ≃ join A B
+joinAlt≃join A B = isoToEquiv (joinAlt-iso-join A B)
+
+joinAlt≡join : (A : Type ℓ) → (B : Type ℓ') → joinAlt A B ≡ join A B
+joinAlt≡join A B = isoToPath (joinAlt-iso-join A B)
+
+
+{-
+  Proof of associativity of the join
+-}
+join-assoc : (A : Type₀) → (B : Type₀) → (C : Type₀)
+             → join (join A B) C ≡ join A (join B C)
+join-assoc A B C = (joinAlt≡join (join A B) C) ⁻¹
+                   ∙ (spanEquivToPath sp3≃sp4) ⁻¹
+                   ∙ (3x3-span.3x3-lemma span) ⁻¹
+                   ∙ (spanEquivToPath sp1≃sp2)
+                   ∙ (joinAlt≡join A (join B C))
+  where
+    -- the meat of the proof is handled by the 3x3 lemma applied to this diagram
+    span : 3x3-span
+    span = record {
+      A00 = A;      A02 = A × B;      A04 = B;
```

Comment:
It would be nice to not have to give all of these fields explicitly. How about making a smarter constructor for 3x3-span that has most of these things implicit?

Reply:
I can do one that only asks for the four homotopies, but that makes it not very convenient — in this instance, Agda will not typecheck the proof anymore. Maybe with some more type annotations in other places it could work.
I can also do one that asks for both functions and homotopies, but then the improvement over the vanilla constructor is not very significant…

Reply:
I'm confused... Why won't Agda check the proof?

Reply:
Well, the functions involved are not so straightforward to extract from the type of the homotopies, since their types only feature compositions of two functions.